### ISSUE TITLE
Variable in scope is not visible after a bang!-operation in CE

### ISSUE THREAD

**Issue by cmeeren** [dotnet/fsharp#13291]
[Repro.zip](https://github.com/dotnet/fsharp/files/8898420/Repro.zip)

At least in the async CE, after a monadic operation (let!/do!/match! etc.), variables in the debugger suddenly seem to go out of scope:

```f#
module Program =

  let f x =
    async {
      do! async.Return ()
      return 1
    }

  [<EntryPoint>]
  let main _argv =
    f 0 |> Async.RunSynchronously |> printfn "%A"
    0
```

![image](https://user-images.githubusercontent.com/7766733/173544086-7b70c3f8-ff18-4bbc-af16-d32dd502456d.png)

![image1](https://user-images.githubusercontent.com/7766733/173544119-adc9da09-e913-411e-b50a-ad000c5113eb.png)

(Screenshots from Rider. Originally [reported there](https://youtrack.jetbrains.com/issue/RIDER-78577), but was redirected here.)

**Comment by vzarytovskii** [comment]
@cmeeren Do you know which compiler version was used? @dsyme was making some changes to how the compiler emits debug points some time ago, I wonder whether it already has those changes or no.

**Comment by cmeeren** [comment]
No idea. A repro solution was attached. You can try that.

**Comment by auduchinok** [comment]
@cmeeren You can look it up here:

<img width="1000" alt="Screenshot 2022-06-14 at 11 51 20" src="https://user-images.githubusercontent.com/3923587/173549179-78bbef56-bd83-4218-b0d0-2e70a2ea08d1.png">

**Comment by cmeeren** [comment]
MSBuild 17.0, SDK 6.0.300.

**Comment by vzarytovskii** [comment]
It's same in VS, I guess it's related to how we store (or rather what we store/not store) in LocalScope/LocalVariable  tables, need to look at which scopes we generate.

**Comment by cmeeren** [comment]
Any chance to have a look at this? It's making debugging rather cumbersome, since I have to insert spurious code everywhere just to be able to break on and inspect the variables I'm after.

**Comment by kerams** [comment]
I agree the situation is not ideal, but if the variable is captured, you will see it under `this` or you can move up the stack a little bit.

**Comment by cmeeren** [comment]
There is nothing under `this`, and nothing further up the stack. Actually no user code at all further up in the stack.

![image](https://user-images.githubusercontent.com/7766733/226286243-dc6889d9-65cc-4848-bd85-2e57ca6edbac.png)

It's the same in VS:

![image](https://user-images.githubusercontent.com/7766733/226286463-f4c9f7f2-2ac0-4753-94ee-5939658a1167.png)

**Comment by vzarytovskii** [comment]
> Any chance to have a look at this? It's making debugging rather cumbersome, since I have to insert spurious code everywhere just to be able to break on and inspect the variables I'm after.

It's not as trivial unfortunately, and will likely require some substantial changes. We'll see when we can plan it.

**Comment by qrzychu** [comment]
Any updates on this? I recently switched jobs and now use F# (love coding in it!), but the debugging side of things are abysmal. 

Sometimes you don't even need a bang operation, normal for loop seems to do the trick also. I guess for release mode it's great that the continuations don't capture the whole scope, but in debug they should capture everything that is available. 

One thing that is probably much harder, is being able to drag the "current line" while debugging up, to rerun things. In C# I could do it whenever I wanted, and with Edit&Continue (plus goodies like Rider highlighting the execution path based on values) the workflow was very nice.

With F# the code is nice, but everything else seems worse, and not by a little bit.
